samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.storage;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.apache.samza.SamzaException;
+import org.apache.samza.container.SamzaContainerMetrics;
+import org.apache.samza.container.TaskInstance;
+import org.apache.samza.metrics.Gauge;
+import org.apache.samza.system.SystemConsumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ *  ContainerStorageManager is a per-container abstraction that instruments
+ *  the restore of per-task partitions.
+ *
+ *  It is responsible for
+ *  a) performing all container-level serializable actions for restore, initializing and shutting down
+ *  taskStorage managers, starting and stopping consumers, etc.
+ *
+ *  b) performing parallelized taskStorageManager restores.
+ *
+ */
+public class ContainerStorageManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ContainerStorageManager.class);
+  private final Map<TaskInstance, TaskStorageManager> taskStorageManagers;
+  private final SamzaContainerMetrics samzaContainerMetrics;
+
+  // Mapping of from storeSystemNames to SystemConsumers
+  private final Map<String, SystemConsumer> systemConsumers;
+
+  // Size of thread-pool to be used for parallel restores
+  private final int parallelRestoreThreadPoolSize;
+
+  // Naming convention to be used for restore threads
+  private static final String RESTORE_THREAD_NAME = "Samza Restore Thread-%d";
+
+  public ContainerStorageManager(Map<TaskInstance, TaskStorageManager> taskStorageManagers,
```

Comment:
for a future PR, we should evaluate re-purposing this as a "RestoreManager" that only handles restores.. thereby stripping restore logic away from the TSM. The TSM wouldn't need access to `SystemConsumer` instances as well

Reply:
yup, agreed.